### DISCUSSION TITLE
Why aren't added types removed when the script exits?

### DISCUSSION THREAD

**Discussion by Worldwidebrine** [PowerShell/PowerShell#18133] [Q&A] [answered]
I execute the script without [dot-sourcing](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scripts?view=powershell-7.2#script-scope-and-dot-sourcing) but the types added by the script exist after script exits.

How to make `Add-Type` scoped?

**Comment by jborean93** [comment, marked as answer]
You can't, `Add-Type` will compile the C# code into a temporary assembly and load that. In dotnet assemblies are loaded process wide and cannot be unloaded once loaded. It's the same for things like binary PowerShell modules.

Technically in the past (dotnet Framework) you could use separate AppDomains but they came with all sorts of limitations, especially around data serialization that made them highly impractical to use.

**Reply by kilasuit** [reply]
Adding to this @jborean93 comment below is the only way you **could** run a script where `Add-Type` is scoped in a way that doesn't pollute the current process with your new  defined types 
> The best way to run in isolation but with real time output is pwsh -File ... which will start a new process with the script you mention. What you want to do really depends on your goals here.

The only addition to this is if you aren't looking for PowerShell versions 6/7 and Windows PowerShell which is version 5, then you'd need to run `powershell -File` instead of `pwsh -file`

**Reply by Worldwidebrine** [reply]
https://github.com/orgs/PowerShell/discussions/18133#discussioncomment-3696643
> You can use this to get the current exe being run
> 
> ```powershell
> (Get-Process -Id $pid).Path
> ```

This is useful. The script can get the PowerShell currently used and start a new process without version detecting.

**Comment by rhubarb-geek-nz** [comment]
A test on MacOS shows that when pswh runs another *.ps1 file it runs it in the same process, but if it runs a normal shell script that runs in a new process. So PowerShell is not following the standard UNIX shell model where all scripts are run in child processes. It first checks to see if it is a PowerShell script and handles it differently to other kinds of executables.

Also for PowerShell to run another *.ps1 file that does not need to have the executable permission set.

If you do want true process isolation then run the child script with "pwsh" within the parent script to force it in a new process.

**Reply by Worldwidebrine** [reply]
But that requires user's extra operation. What is the best way to let script isolate itself?
`Start-Job` redirects output but I want real-time output. I tried `Start-Process` but I don't know how to get the current interpreter.
And is it possible that PowerShell would change it?

**Reply by jborean93** [reply]
The best way to run in isolation but with real time output is `pwsh -File ...` which will start a new process with the script you mention. What you want to do really depends on your goals here.

**Reply by rhubarb-geek-nz** [reply]
If the script needs to self isolate then it should run either itself or another script with pwsh. This could be done with a command line flag that if it is not set then the script runs itself again with pwsh and the original arguments and the special flag set. This can be done by the script itself.

**Reply by Worldwidebrine** [reply]
`pwsh` is a program not a command, right? I don't think that's a good idea. `pwsh` doesn't works on Windows that has only preinstalled PowerShell.
Can script get the interpreter that is currently interpreting the script?

**Reply by rhubarb-geek-nz** [reply]
If you put try/catch around the invoking pwsh you will get an exception if the program is not found, then you could run PowerShell.exe. This assumes of course that you have written a PowerShell script for the original PowerShell. This GitHub project is for PowerShell Core which is the cross-platform solution and the program is called pwsh. To determine the version of PowerShell that is running use $PSVersionTable.PSVersion

**Reply by Worldwidebrine** [reply]
You mean there is no way for script to get the interpreter that is currently interpreting the script?

**Reply by rhubarb-geek-nz** [reply]
To determine the version of PowerShell that is running use $PSVersionTable.PSVersion

**Reply by rhubarb-geek-nz** [reply]
Write-Host $Env:_ 
/usr/local/bin/pwsh

**Reply by Worldwidebrine** [reply]
`$Env:_` is null

**Reply by rhubarb-geek-nz** [reply]
Use $PSVersionTable.PSVersion
If major is six or higher then use pwsh, else use PowerShell.exe. Alternatively if your script can run using legacy PowerShell then always use PowerShell.exe on Windows, and pwsh on Linux/MacOS. If your program won't run on legacy PowerShell then there is no point worrying about it.

**Reply by jborean93** [reply]
You can use this to get the current exe being run

```powershell
(Get-Process -Id $pid).Path
```
